plugins/wallet_plugin/se_wallet.cpp

```diff
@@ -0,0 +1,372 @@
+      macos_user_auth(auth_callback, &prom, CFSTR("remove a key from your EOSIO wallet"));
+      if(!fut.get())
+         FC_THROW_EXCEPTION(chain::wallet_invalid_password_exception, "Local user authentication failed");
+
+      CFDictionaryRef deleteDic = CFDictionaryCreate(nullptr, (const void**)&kSecValueRef, (const void**)&it->second, 1, &kCFTypeDictionaryKeyCallBacks, &kCFTypeDictionaryValueCallBacks);
+
+      OSStatus ret = SecItemDelete(deleteDic);
+      CFRelease(deleteDic);
+
+      if(ret)
+         FC_THROW_EXCEPTION(chain::wallet_exception, "Failed to getremove key from Secure Enclave");
+
+      CFRelease(it->second);
+      _keys.erase(it);
+
+      return true;
+   }
+
+   ~se_wallet_impl() {
+      for(auto& k : _keys)
+         CFRelease(k.second);
+   }
+
+   map<public_key_type,SecKeyRef> _keys;
+   fc::ec_key key = EC_KEY_new_by_curve_name(NID_X9_62_prime256v1);
+   bool locked = true;
+};
+
+static void check_signed() {
+   OSStatus is_valid{0};
+   pid_t pid = getpid();
+   SecCodeRef code = nullptr;
+   CFNumberRef pidnumber = CFNumberCreate(kCFAllocatorDefault, kCFNumberSInt32Type, &pid);
+   CFDictionaryRef piddict = CFDictionaryCreate(kCFAllocatorDefault, (const void**)&kSecGuestAttributePid, (const void**)&pidnumber, 1, nullptr, nullptr);
+   if(!SecCodeCopyGuestWithAttributes(nullptr, piddict, kSecCSDefaultFlags, &code)) {
+      is_valid = SecCodeCheckValidity(code, kSecCSDefaultFlags, 0);
+      CFRelease(code);
+   }
+   CFRelease(piddict);
+   CFRelease(pidnumber);
+
+   if(is_valid != errSecSuccess) {
+      wlog("Application does not have a valid signature; Secure Enclave support disabled");
+      FC_THROW("");
+   }
+}
+
+}
+
+se_wallet::se_wallet() : my(new detail::se_wallet_impl()) {
+   detail::check_signed();
+
+   //How to figure out of SE is available?!
+   char model[256];
+   size_t model_size = sizeof(model);
+   if(sysctlbyname("hw.model", model, &model_size, nullptr, 0) == 0) {
+      if(strncmp(model, "iMacPro", strlen("iMacPro")) == 0) {
+         my->populate_existing_keys();
+         return;
+      }
+      unsigned int major, minor;
+      if(sscanf(model, "MacBookPro%u,%u", &major, &minor) == 2) {
+         if(major >= 13 && minor >= 2) {
+            my->populate_existing_keys();
+            return;
+         }
+      }
+   }
+
+   FC_THROW("Secure Enclave not supported on this hardware");
+}
+
+se_wallet::~se_wallet() {
+}
+
+private_key_type se_wallet::get_private_key(public_key_type pubkey) const {
+   FC_THROW_EXCEPTION(chain::wallet_exception, "Obtaining private key for a key stored in Secure Enclave is impossible");
+}
+
+bool se_wallet::is_locked() const {
+   return my->locked;
+}
+void se_wallet::lock() {
+   FC_ASSERT(!is_locked());
+   my->locked = true;
+}
+
+void se_wallet::unlock(string password) {
+   promise<bool> prom;
+   future<bool> fut = prom.get_future();
+   macos_user_auth(detail::auth_callback, &prom, CFSTR("unlock your EOSIO wallet"));
+   if(!fut.get())
```

Comment:
This will block execution until the user interacts with the GUI. Lousy for keosd but still manageable I think. It’s probably more of a problem if using wallet inside of nodeos